### PR TITLE
Vm snapshot

### PR DESCRIPTION
**Pull request by domendobnikar**
As a user I would like to be able to create, list, delete and clone from VM level snapshots using “hypercore” collection.
Cloning is done in a separate PR. https://github.com/ScaleComputing/HyperCoreAnsibleCollection/pull/219
- [x] Snapshots
- [x] Snapshot info
- [x] Docs
- [x] Split PR (clone)
---------------------------------------
Testing
- [x] Integration
- [x] Units

### CODE REVIEW

File: plugins/module_utils/vm.py [hunk 1]
```diff
@@ -307,8 +310,11 @@ def create_clone_vm_payload(
         *,
         preserve_mac_address,
         source_nics,
+        snapshot_label,
     ):
-        data = {"template": {}}
+        data = dict(template=dict())
+        if snapshot_label:
+            data["snapUUID"] = snapshot_label
```
Comment (justinc1) [inline plugins/module_utils/vm.py:317]: I think you should split this PR into two smaller PRs - one is about vm_snapshot module, how to create and delete them, second is about how to use VM snapshots - clone new VMs from them etc.

The marked line,s I also needed same thing in #218. I just assigned also you to review.

Comment (domendobnikar) [reply plugins/module_utils/vm.py:317]: Good idea, will do 👍

**Comment by domendobnikar** [comment]
Integration tests:
https://github.com/ScaleComputing/HyperCoreAnsibleCollection/actions/runs/4829318439

**Review by justinc1 — DISMISSED** [review]
Just trivial suggestions.

File: plugins/modules/vm_snapshot.py [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+
+
+def calculate_date(days: Optional[int]) -> Optional[datetime.date]:
+    if days is None or days == 0:
+        return None
+    return datetime.datetime.today() + datetime.timedelta(days=days)
+
+
+def check_parameters(module: AnsibleModule) -> AnsibleModule:
+    module.params["retain_for"] = calculate_date(module.params["retain_for"])
```
[the remark this answers is not in the document]
Comment (domendobnikar) [reply plugins/modules/vm_snapshot.py]: I moved the entire logic to module utils, this way we calculate date for an object of class `VMSnapshot` and never for dictionary.

[the remark this answers is not in the document]
Comment (justinc1) [reply plugins/modules/vm_snapshot.py]: I wanted to say, do not replace value of `module.params["retain_for"]`. If calculated timestamp needs to be stored `into module.params`, than at least store it with a different name - like `module.params["retain_until_timestamp"]`.

Code is harder to undestand if value is first `3` (3 days) , then it just changes to a `1683635981`.

**Review by justinc1 — CHANGES_REQUESTED** [review]
I would really like that input parameter `retain_for` value is not replaced. Just store computed value with a new name, and use new name later.

**Comment by domendobnikar** [comment]
It's not being replaced anymore.
I also moved the `calculate_date `logic to utils, since it makes more sense to do it once during `from_ansible` when making an object and have it stored within the object instead of dictionary.
So now the `module.params` is not being altered at all, the calculated time stamp value is stored inside the `VMsnapshot` object instead.

**Comment by justinc1** [comment]
Sorry @domendobnikar I complained a second time about "do not replace value of module.params["retain_for"]" - I guess I was looking at old code. my bad :/

**Pull request merged**